File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    
    int helper(int indx,int n,vector<int>&nums,vector<int>&dp)
    {
        if(indx>=n-1)
        {
            return 0;
        }
        if(dp[indx]!=-1)
            return dp[indx];
        int minJmp=1e9;
        for(int i=indx+1;i<=indx+nums[indx];i++)
        {
            minJmp=min(minJmp,1+helper(i,n,nums,dp));
        }
        return dp[indx]=minJmp;
    }
    int jump(vector<int>& nums) {
        vector<int>dp(nums.size(),-1);
        int n=nums.size();
        cout<<n<<"\n";
        return helper(0,n,nums,dp);
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        if(nums.length==1)
            return 0;
        int reach=nums[0],max_reach=nums[0],step=1;
        for(int i=1;i<nums.length;i++)
        {
            if(reach==0)
            {
                step++;
                reach=max_reach;
            }
            reach--;
            max_reach=Math.max(max_reach-1,nums[i]);
        }
        return step;
    }
}